internal/rm/health.go

```diff
@@ -17,189 +17,460 @@
+
+	spec "github.com/NVIDIA/k8s-device-plugin/api/config/v1"
 )
 
 const (
-	// envDisableHealthChecks defines the environment variable that is checked to determine whether healthchecks
-	// should be disabled. If this envvar is set to "all" or contains the string "xids", healthchecks are
-	// disabled entirely. If set, the envvar is treated as a comma-separated list of Xids to ignore. Note that
-	// this is in addition to the Application errors that are already ignored.
+	// envDisableHealthChecks defines the environment variable that is
+	// checked to determine whether healthchecks should be disabled. If
+	// this envvar is set to "all" or contains the string "xids",
+	// healthchecks are disabled entirely. If set, the envvar is treated
+	// as a comma-separated list of Xids to ignore. Note that this is in
+	// addition to the Application errors that are already ignored.
 	envDisableHealthChecks = "DP_DISABLE_HEALTHCHECKS"
-	// envEnableHealthChecks defines the environment variable that is checked to
-	// determine which XIDs should be explicitly enabled. XIDs specified here
-	// override the ones specified in the `DP_DISABLE_HEALTHCHECKS`.
-	// Note that this also allows individual XIDs to be selected when ALL XIDs
-	// are disabled.
+	// envEnableHealthChecks defines the environment variable that is
+	// checked to determine which XIDs should be explicitly enabled. XIDs
+	// specified here override the ones specified in the
+	// `DP_DISABLE_HEALTHCHECKS`. Note that this also allows individual
+	// XIDs to be selected when ALL XIDs are disabled.
 	envEnableHealthChecks = "DP_ENABLE_HEALTHCHECKS"
 )
 
-// CheckHealth performs health checks on a set of devices, writing to the 'unhealthy' channel with any unhealthy devices
-func (r *nvmlResourceManager) checkHealth(stop <-chan interface{}, devices Devices, unhealthy chan<- *Device) error {
-	xids := getDisabledHealthCheckXids()
-	if xids.IsAllDisabled() {
+// HealthProvider manages GPU device health monitoring with lifecycle
+// control.
+type HealthProvider interface {
+	// Start initiates health monitoring. Blocks until initial setup
+	// completes. Returns error if health monitoring cannot be started.
+	Start(context.Context) error
+
+	// Stop gracefully shuts down health monitoring and waits for all
+	// goroutines to complete. Safe to call multiple times.
+	Stop()
+
+	// Health returns a read-only channel that receives devices that
+	// have become unhealthy.
+	Health() <-chan *Device
+}
+
+// nvmlHealthProvider implements HealthProvider using NVML event
+// monitoring. This is a refactoring of the existing checkHealth logic
+// with proper lifecycle management.
+type nvmlHealthProvider struct {
+	// Configuration
+	nvml    nvml.Interface
+	config  *spec.Config
+	devices Devices
+
+	// NVML resources
+	eventSet nvml.EventSet
+
+	// Lifecycle management
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	// State guards
+	mu      sync.Mutex
```

Comment:
We could use an `IsA` relationship to simplify taking and releasing the lock:

```suggestion
	sync.Mutex
```